core/src/main/java/com/linecorp/armeria/internal/common/LeakTracingRequestContextStorage.java

```diff
@@ -0,0 +1,134 @@
+
+import static java.lang.Thread.currentThread;
+import static java.util.Objects.requireNonNull;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.ClientRequestContextWrapper;
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.RequestContextStorage;
+import com.linecorp.armeria.common.RequestContextWrapper;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.common.util.Sampler;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.ServiceRequestContextWrapper;
+
+/**
+ * A {@link RequestContextStorage} which keeps track of {@link RequestContext}s, reporting pushed thread
+ * information if a {@link RequestContext} is leaked.
+ */
+@UnstableApi
+final class LeakTracingRequestContextStorage implements RequestContextStorage {
+
+    private final RequestContextStorage delegate;
+    private final Sampler<? super RequestContext> sampler;
+
+    /**
+     * Creates a new instance.
+     * @param delegate the underlying {@link RequestContextStorage} that stores {@link RequestContext}
+     * @param sampler the {@link Sampler} that determines whether to retain the stacktrace of the context leaks
+     */
+    LeakTracingRequestContextStorage(RequestContextStorage delegate,
+                                            Sampler<? super RequestContext> sampler) {
+        this.delegate = requireNonNull(delegate, "delegate");
+        this.sampler = requireNonNull(sampler, "sampler");
+    }
+
+    @Nullable
+    @Override
+    public <T extends RequestContext> T push(RequestContext toPush) {
+        requireNonNull(toPush, "toPush");
+        if (sampler.isSampled(toPush)) {
+            return delegate.push(warpRequestContext(toPush));
+        }
+        return delegate.push(toPush);
+    }
+
+    @Override
+    public void pop(RequestContext current, @Nullable RequestContext toRestore) {
+        requireNonNull(current, "current");
+        delegate.pop(current, toRestore);
+    }
+
+    @Nullable
+    @Override
+    public <T extends RequestContext> T currentOrNull() {
+        return delegate.currentOrNull();
+    }
+
+    private static RequestContextWrapper<?> warpRequestContext(RequestContext ctx) {
+        return ctx instanceof ClientRequestContext ?
+               new TraceableClientRequestContext((ClientRequestContext) ctx)
+               : new TraceableServiceRequestContext((ServiceRequestContext) ctx);
+    }
+
+    private static final class TraceableClientRequestContext extends ClientRequestContextWrapper {
+
+        private final PendingRequestContextStackTrace stackTrace;
+
+        private TraceableClientRequestContext(ClientRequestContext delegate) {
+            super(delegate);
+            stackTrace = new PendingRequestContextStackTrace();
```

Comment:
How about using `StackTraceElement[]` because
- an exception is not raised yet (we don't need to create an instance of an exception that might not be raised in the future.)
- all we need in `toString` is the stack trace?

If we using `StackTraceElement[]`, we could probably do:
```java
private TraceableClientRequestContext(ServiceRequestContext delegate) {
    super(delegate);
    stackTrace = currentThread().getStackTrace();
}

@Override
public String toString() {
    final StringBuilder builder = new StringBuilder(10000); // What's the sensible default?
    builder.append(super.toString());
    builder.append(System.lineSeparator());
    // Start with 1 not to print .getStackTrace(Thread.java:1610)
    for (int i = 1; i < stackTrace.length; i++) {
        builder.append("\tat ").append(stackTrace[i]).append(System.lineSeparator());
    }
    return builder.toString();
}

```

Reply:
JFYI, `StackTraceWriter` uses 512 for the initial buffer. https://github.com/line/armeria/blob/5b384fbe27e7e6f9225d6db91cbb684d09dfbb5e/core/src/main/java/com/linecorp/armeria/common/util/Exceptions.java#L314-L316